File: ai_trader/app/engine/indicators.py
"""Индикаторы и признаки рынка по 5-минутным свечам (без внешних библиотек)."""
from statistics import median
from typing import Dict, List, Optional


def ema(values: List[float], n: int) -> List[float]:
    if not values:
        return []
    k = 2 / (n + 1)
    out = [values[0]]
    for v in values[1:]:
        out.append(v * k + out[-1] * (1 - k))
    return out


def true_ranges(h: List[float], l: List[float], c: List[float]) -> List[float]:
    tr = [h[0] - l[0]]
    for i in range(1, len(c)):
        tr.append(max(h[i] - l[i], abs(h[i] - c[i - 1]), abs(l[i] - c[i - 1])))
    return tr


def wilder(values: List[float], n: int) -> List[float]:
    out, acc = [], None
    for i, v in enumerate(values):
        if i < n:
            acc = (acc or 0) + v
            out.append(acc / (i + 1))
        else:
            out.append((out[-1] * (n - 1) + v) / n)
    return out


def rsi(c: List[float], n: int = 14) -> float:
    gains = [max(c[i] - c[i - 1], 0) for i in range(1, len(c))]
    losses = [max(c[i - 1] - c[i], 0) for i in range(1, len(c))]
    g, lo = wilder(gains, n)[-1], wilder(losses, n)[-1]
    return 100.0 if lo == 0 else 100 - 100 / (1 + g / lo)


def adx(h: List[float], l: List[float], c: List[float], n: int = 14) -> float:
    plus_dm, minus_dm = [0.0], [0.0]
    for i in range(1, len(c)):
        up, down = h[i] - h[i - 1], l[i - 1] - l[i]
        plus_dm.append(up if up > down and up > 0 else 0.0)
        minus_dm.append(down if down > up and down > 0 else 0.0)
    tr = wilder(true_ranges(h, l, c), n)
    pdi = [100 * p / t if t else 0 for p, t in zip(wilder(plus_dm, n), tr)]
    mdi = [100 * m / t if t else 0 for m, t in zip(wilder(minus_dm, n), tr)]
    dx = [100 * abs(p - m) / (p + m) if p + m else 0 for p, m in zip(pdi, mdi)]
    return wilder(dx, n)[-1]


def features(klines: List[list], price: Optional[float] = None) -> Optional[Dict[str, float]]:
    """klines: старые -> новые, [ts, open, high, low, close, volume, turnover]. Последняя свеча может быть незакрытой."""
    if len(klines) < 210:
        return None
    o = [float(k[1]) for k in klines]
    h = [float(k[2]) for k in klines]
    l = [float(k[3]) for k in klines]
    c = [float(k[4]) for k in klines]
    vol = [float(k[5]) for k in klines]
    turn = [float(k[6]) for k in klines]
    price = price or c[-1]

    e20, e50, e200 = ema(c, 20), ema(c, 50), ema(c, 200)
    tr = true_ranges(h, l, c)
    atr_series = wilder(tr, 14)
    atr = atr_series[-1]
    atr_pcts = [a / cl for a, cl in zip(atr_series[-200:], c[-200:])]
    w = c[-20:]
    mean20 = sum(w) / 20
    std20 = (sum((x - mean20) ** 2 for x in w) / 20) ** 0.5
    v48 = sum(vol[-48:])
    vwap4h = sum(turn[-48:]) / v48 if v48 else price

    return {
        "price": price,
        "ema20": e20[-1], "ema50": e50[-1], "ema200": e200[-1],
        "atr": atr, "atr_pct": atr / price * 100,
        "atr_pct_median": median(atr_pcts) * 100,
        "adx": adx(h, l, c),
        "rsi": rsi(c),
        "bb_width_pct": 4 * std20 / mean20 * 100,
        "ret_1h": (price / c[-13] - 1) * 100,
        "ret_4h": (price / c[-49] - 1) * 100,
        "ret_24h": (price / c[-289] - 1) * 100 if len(c) >= 289 else 0.0,
        "vwap_4h": vwap4h,
        "dist_vwap_atr": (price - vwap4h) / atr if atr else 0.0,
        # для точки входа трендовой стратегии (по последней ЗАКРЫТОЙ свече)
        "last_closed_ts": float(klines[-2][0]),
        "c1": c[-2], "o1": o[-2], "h2": h[-3], "l2": l[-3],
        "low_5": min(l[-7:-1]), "high_5": max(h[-7:-1]),
        "low_10": min(l[-12:-1]), "high_10": max(h[-12:-1]),
        "ema20_1": e20[-2],
    }


REGIMES = ("trend_up", "trend_down", "range", "high_volatility")


def rule_regime(f: Dict[str, float]) -> str:
    """Запасной классификатор режима, если ИИ недоступен."""
    if f["atr_pct"] > 2.2 * f["atr_pct_median"]:
        return "high_volatility"
    if f["adx"] >= 23 and f["ema50"] > f["ema200"] and f["price"] > f["ema50"]:
        return "trend_up"
    if f["adx"] >= 23 and f["ema50"] < f["ema200"] and f["price"] < f["ema50"]:
        return "trend_down"
    return "range"

File: izidva/src/Engine/Indicators.php
<?php
declare(strict_types=1);

namespace App\Engine;

/** Индикаторы и признаки рынка по 5-минутным свечам. */
final class Indicators
{
    public const REGIMES = ['trend_up', 'trend_down', 'range', 'high_volatility'];

    public static function ema(array $v, int $n): array
    {
        if (!$v) {
            return [];
        }
        $k = 2 / ($n + 1);
        $out = [$v[0]];
        for ($i = 1, $c = count($v); $i < $c; $i++) {
            $out[] = $v[$i] * $k + $out[$i - 1] * (1 - $k);
        }
        return $out;
    }

    public static function trueRanges(array $h, array $l, array $c): array
    {
        $tr = [$h[0] - $l[0]];
        for ($i = 1, $n = count($c); $i < $n; $i++) {
            $tr[] = max($h[$i] - $l[$i], abs($h[$i] - $c[$i - 1]), abs($l[$i] - $c[$i - 1]));
        }
        return $tr;
    }

    public static function wilder(array $v, int $n): array
    {
        $out = [];
        $acc = 0.0;
        foreach ($v as $i => $x) {
            if ($i < $n) {
                $acc += $x;
                $out[] = $acc / ($i + 1);
            } else {
                $out[] = ($out[$i - 1] * ($n - 1) + $x) / $n;
            }
        }
        return $out;
    }

    public static function rsi(array $c, int $n = 14): float
    {
        $g = $lo = [];
        for ($i = 1, $m = count($c); $i < $m; $i++) {
            $g[] = max($c[$i] - $c[$i - 1], 0);
            $lo[] = max($c[$i - 1] - $c[$i], 0);
        }
        $wg = self::wilder($g, $n);
        $wl = self::wilder($lo, $n);
        $ag = (float)end($wg);
        $al = (float)end($wl);
        return $al == 0.0 ? 100.0 : 100 - 100 / (1 + $ag / $al);
    }

    public static function adx(array $h, array $l, array $c, int $n = 14): float
    {
        $pdm = [0.0];
        $mdm = [0.0];
        for ($i = 1, $m = count($c); $i < $m; $i++) {
            $up = $h[$i] - $h[$i - 1];
            $down = $l[$i - 1] - $l[$i];
            $pdm[] = ($up > $down && $up > 0) ? $up : 0.0;
            $mdm[] = ($down > $up && $down > 0) ? $down : 0.0;
        }
        $tr = self::wilder(self::trueRanges($h, $l, $c), $n);
        $wp = self::wilder($pdm, $n);
        $wm = self::wilder($mdm, $n);
        $dx = [];
        foreach ($tr as $i => $t) {
            $p = $t ? 100 * $wp[$i] / $t : 0;
            $mi = $t ? 100 * $wm[$i] / $t : 0;
            $dx[] = ($p + $mi) ? 100 * abs($p - $mi) / ($p + $mi) : 0;
        }
        $w = self::wilder($dx, $n);
        return (float)end($w);
    }

    /**
     * @param array $k свечи от старых к новым: [ts, open, high, low, close, volume, turnover]
     *                 (последняя может быть незакрытой)
     */
    public static function features(array $k, ?float $price = null): ?array
    {
        $n = count($k);
        if ($n < 210) {
            return null;
        }
        $o = array_map(fn($x) => (float)$x[1], $k);
        $h = array_map(fn($x) => (float)$x[2], $k);
        $l = array_map(fn($x) => (float)$x[3], $k);
        $c = array_map(fn($x) => (float)$x[4], $k);
        $vol = array_map(fn($x) => (float)$x[5], $k);
        $turn = array_map(fn($x) => (float)$x[6], $k);
        $price ??= $c[$n - 1];

        $e20 = self::ema($c, 20);
        $e50 = self::ema($c, 50);
        $e200 = self::ema($c, 200);
        $atrS = self::wilder(self::trueRanges($h, $l, $c), 14);
        $atr = $atrS[$n - 1];
        $atrPcts = [];
        for ($i = max(0, $n - 200); $i < $n; $i++) {
            $atrPcts[] = $atrS[$i] / $c[$i];
        }
        sort($atrPcts);
        $mid = intdiv(count($atrPcts), 2);
        $median = count($atrPcts) % 2 ? $atrPcts[$mid] : ($atrPcts[$mid - 1] + $atrPcts[$mid]) / 2;
        $w = array_slice($c, -20);
        $mean = array_sum($w) / 20;
        $std = sqrt(array_sum(array_map(fn($x) => ($x - $mean) ** 2, $w)) / 20);
        $v48 = array_sum(array_slice($vol, -48));
        $vwap = $v48 ? array_sum(array_slice($turn, -48)) / $v48 : $price;

        return [
            'price' => $price,
            'ema20' => $e20[$n - 1], 'ema50' => $e50[$n - 1], 'ema200' => $e200[$n - 1],
            'atr' => $atr, 'atr_pct' => $atr / $price * 100, 'atr_pct_median' => $median * 100,
            'adx' => self::adx($h, $l, $c), 'rsi' => self::rsi($c),
            'bb_width_pct' => 4 * $std / $mean * 100,
            'ret_1h' => ($price / $c[$n - 13] - 1) * 100,
            'ret_4h' => ($price / $c[$n - 49] - 1) * 100,
            'ret_24h' => $n >= 289 ? ($price / $c[$n - 289] - 1) * 100 : 0.0,
            'vwap_4h' => $vwap,
            'dist_vwap_atr' => $atr ? ($price - $vwap) / $atr : 0.0,
            // точка входа — по последней закрытой свече
            'last_closed_ts' => (float)$k[$n - 2][0],
            'c1' => $c[$n - 2], 'o1' => $o[$n - 2], 'h2' => $h[$n - 3], 'l2' => $l[$n - 3],
            'low_5' => min(array_slice($l, -7, 6)), 'high_5' => max(array_slice($h, -7, 6)),
            'low_10' => min(array_slice($l, -12, 11)), 'high_10' => max(array_slice($h, -12, 11)),
            'ema20_1' => $e20[$n - 2],
        ];
    }

    public static function ruleRegime(array $f): string
    {
        if ($f['atr_pct'] > 2.2 * $f['atr_pct_median']) {
            return 'high_volatility';
        }
        if ($f['adx'] >= 23 && $f['ema50'] > $f['ema200'] && $f['price'] > $f['ema50']) {
            return 'trend_up';
        }
        if ($f['adx'] >= 23 && $f['ema50'] < $f['ema200'] && $f['price'] < $f['ema50']) {
            return 'trend_down';
        }
        return 'range';
    }
}
